Feature/32 product table crud

stishchenko opened the pull request:
CRUD for Product table with tests

Tishchenko666 committed:
Class Product -> entity of product table was created; has constructors,getters,setters.
Enum ProductState -> states of Product.
Interface ProductDao -> service to connect to DB; searching by id, saving, updating, deleting object.
Class ProductService -> service uses dao to connect to DB; searching by id, saving, updating, deleting object.

Tishchenko666 committed:
Implementation of ProductDao.
Creation of ProductServiceTest fot ProductService.

Tishchenko666 committed:
Product -> add version field.
ProductDao -> in method delete() replace arg Product with Integer id.
ProductService -> get ProductDao in constructor, validate args before using dao: methods readById() and deleteProduct() have id not null and higher than 0 - return null; methods createProduct() and updateProduct() have product not null, its required fields not null - throw ValidationException.
ValidationException -> custom exception.
------
ProductServiceTest -> test ProductService using test implementation of ProductDao.
ProductDaoTestImpl -> test implementation of ProductDao.

rdovgan commented on src/main/java/team/asd/dao/ProductDao.java:
```diff
@@ -0,0 +1,14 @@
+
+public interface ProductDao {
+
+    Product findById(Integer id);
```
change method to `readById`

rdovgan commented on src/main/java/team/asd/service/ProductService.java:
```diff
@@ -0,0 +1,49 @@
+        if (product == null) {
+            return null;
+        }
+        if (ObjectUtils.anyNull(product.getSupplierId(), product.getName(), product.getCurrency())) {
```
you can extract validation to a separate method

rdovgan commented on src/main/java/team/asd/service/ProductService.java:
```diff
@@ -0,0 +1,49 @@
+        if (product == null) {
+            return null;
+        }
+        if (ObjectUtils.anyNull(product.getProductId(), product.getSupplierId(), product.getName(), product.getCurrency())) {
```
in `update` method any field can be `null` except `id`

rdovgan commented on src/test/java/team/asd/dao/ProductDaoTestImpl.java:
```diff
@@ -0,0 +1,36 @@
+
+    private Product expectedProduct = new Product();
+
+    public Product getExpectedProduct() {
```
move this method to new class `ProductDataTest` class, because it has no relation to DAO service implementation

rdovgan commented on src/test/java/team/asd/service/ProductServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+        assertNull(productService.deleteProduct(-1));
+        assertEquals(product.getProductId(), productService.deleteProduct(1));
+    }
+}
```
need to add empty line

Tishchenko666 committed:
Use Lombok to Product

rdovgan commented on src/main/java/team/asd/entity/Product.java:
```diff
@@ -0,0 +1,52 @@
+@AllArgsConstructor
+public class Product {
+
+    private Integer productId;
```
Change to `id`

Tishchenko666 committed:
#32
Product -> rename productId in id.
ProductDao -> rename findById in readById.
ProductDataTest -> was created to get product for tests.
ProductDaoTestImpl -> replace expectedProduct from it to ProductDataTest.
ProductService -> extract validation to sepaarte methods validId() and validProduct().
ProductServiceTest -> use ProductDataTest to assertEquals.

Tishchenko666 committed:
ProductServiceTest -> replace null test to exception test.
